src/RF24LogAbstracts/BaseHandler.h

```diff
@@ -0,0 +1,114 @@
+/**
+ * @file RF24LogAbstracts/BaseHandler.h
+ * @brief inherent declarations for all handlers.
+ * @date Created 2 Oct 2020
+ * @author Witold Markowski (wmarkow)
+ * @copyright Copyright (C)
+ *      2020        Witold Markowski (wmarkow)
+ *      2021        Brendan Doherty (2bndy5) <br>
+ * This General Public License does not permit incorporating your program into
+ * proprietary programs.  If your program is a subroutine library, you may
+ * consider it more useful to permit linking proprietary applications with the
+ * library.  If this is what you want to do, use the GNU Lesser General
+ * Public License instead of this License.
+ */
+
+#ifndef SRC_RF24LOGABSTRACTS_BASEHANDLER_H_
+#define SRC_RF24LOGABSTRACTS_BASEHANDLER_H_
+
+#if defined(ARDUINO_ARCH_AVR)
+#include <WString.h> // __FlashStringHelper
+#else
+#include <string.h>
+#endif
+#include <stdint.h>
+#include <stdarg.h>
+
+/** @brief Change The Delimiter character used in the header prefix of log messages. */
+#if !defined(RF24LOG_DELIMITER)
+#define RF24LOG_DELIMITER ';'
+#endif
+
+// for some reason, these need to be declared here and
+// not on any higher level of the project
+// #define RF24LOG_NO_TIMESTAMP
+// #define RF24LOG_NO_EOL
+// #define RF24LOG_SHORT_DESC
+// #define RF24LOG_TERSE_DESC
+// #define RF24LOG_TAB_SIZE 4
+
+#ifdef DOXYGEN_FORCED
+/** @brief macro (when defined) disables line feeds at the end of all log messages. */
+#define RF24LOG_NO_EOL
+
+/** @brief macro (when defined) disables timestamp at the start of all log messages. */
+#define RF24LOG_NO_TIMESTAMP
+
+/**
+ * @brief macro (when defined) uses a short description of the log levels.
+ *
+ * Descriptions with this macro defined look like
+ * @code
+ * " ERR ", "INFO1"
+ * "WARN ", "INFO7"
+ * "INFO ", "L   7"
+ * " DBG ", "L 113"
+ * @endcode
+ */
+#define RF24LOG_SHORT_DESC
+
+/**
+ * @brief macro (when defined) uses a terse (very short) description of the log levels.
+ *
+ * Descriptions with this macro defined look like
+ * @code
+ * " E ", " I1"
+ * " W ", " I7"
+ * " I ", "  7"
+ * "DB ", "113"
+ * @endcode
+ */
+#define RF24LOG_TERSE_DESC
+
+/**
+ * @brief macro (when defined) relaces all `\t` characters with a specified quantity of spaces
+ *
+ * When this macro is not defined the `\t` characters may have in consistent number of spaces between platforms.
+ */
+#define RF24LOG_TAB_SIZE 8
+#endif
```

Comment:
his should be moved to the first real implementation of the handler, which in our case is ```Handler.h```.

Reply:
These macro definitions are for docs only (notice the `#ifdef DOXYGEN_FORCED`). My original intention was to let the user define the macro(s) they want in the sketch... but as you can see by the line comments above, this was the highest level on which the macros have any affect.

Reply:
Does using a config.h file that the abstract handler includes make sense for macro definitions/docs?

Reply:
TBH I have no idea how to handle correctly macro definitions with the possibility of overriding them in the final C++ project. 

>My original intention was to let the user define the macro(s) they want in the sketch... 

That would be good to have but how to do it?

>Does using a config.h file that the abstract handler includes make sense for macro definitions/docs?

Hm, will that work? The **config.h** or even **rf24log_config.h** (to distinguish between other config files of a different libraries) should be provided then by the final application. What about our Examples? Does it mean that each example folder should contain a config.h? It would be good to have something like: if the user doesn't provide his config then use a default one from the inside of RF24Log. I don't know if that's possible.

Reply:
No like _RF24LogConfig.h_ . It would be **the** place for any user to go and uncomment or comment out a macro of their choosing. No need to keep a copy in every example folder. 

>>My original intention was to let the user define the macro(s) they want in the sketch...
>That would be good to have but how to do it?

I'm not sure we can because of the singleton or because of the src folder.

1. Arduino IDE complies everything it finds inside a lib's src folder (which is why `OStreamLogger` is completely blocked out to the IDE using `#ifndef ARDUINO ... #endif`).
2. The singleton does need to know about any of these macros. The highest level that correctly affects all loggers is the Base handler.

When I tried to use the macros in the examples, it didn't work. I even put the macros before all other RF24Log include statements; still no avail. So I'm not convinced a RF24LogConfig.h file is going to work any better for pre-defining macros in the example. There's also the possibility to use our own namespace for macros, but IDK if that would work any better.

Reply:
@wmarkow now that you know that these macro defs are for docs only and that defining the macros in this file works just as well as defining them in a file containing child of the base handler, Can I leave these macro defs here in the base handler class' file?

Later , we can re-visit customizing macros on the application's scope rather than the lib's source files.

Reply:
Ok, let's leave it for now as is. We can improve/fix it later.